### DISCUSSION TITLE
Is there a way to limit edge middleware to only run on pages?

### DISCUSSION THREAD

**Discussion by DerekKHarris** [vercel/next.js#39435] [Help] [answered]
I recently implemented middleware to set a cookie anytime someone visits the site using a refid.  I'm getting slammed by Vercel on Edge Middleware Invocations and as of yesterday have gone past my 1,000,000 usages.  This is running on every asset even though I only want it to be invoked on pages.  How would I limit this to only invoke on pages?  As you can see in the code, I've tried everything I could find online yet haven't been able to resolve the issue.

`import { NextResponse } from "next/server";

const PUBLIC_FILE = /\.(.*)$/;

export function middleware(req) {
	const { pathname } = req.nextUrl;

	if (
		pathname.startsWith("/_next") || // exclude Next.js internals
		pathname.startsWith("/api") || //  exclude all API routes
		pathname.startsWith("/static") || // exclude static files
		PUBLIC_FILE.test(pathname) // exclude all files in the public folder
	) {
		return NextResponse.next();
	} else {
		const response = NextResponse.next();
		if (req.nextUrl.searchParams.get("refid")) {
			response.cookies.set(
				"affiliate_ref_id",
				req.nextUrl.searchParams.get("refid"),
				{
					maxAge: 2628000,
				}
			);
		}

		return response;
	}
}

export const config = {
	matcher: ["/:path*"],
};
`

**Comment by dasblitz** [comment, marked as answer]
I was looking for the same, for the same reasons :) I found a PR to improve the docs on `config.matcher` explaining you can also exclude folders (paths): 

https://github.com/vercel/next.js/pull/40180/files#:~:text=%7D-,export%20const%20config%20%3D%20%7B,%7D,-13

I've just tried this locally and it seems to work:

```javascript
// match files that don't start with /public, /static etc...
export const config = {
  matcher: ['/((?!public|static|api|_next).*)'],
};
```

 let me know if that helps :)

**Reply by DerekKHarris** [reply]
This is the way :) thanks!
